src/lib/mdns/DiscoveryManager.cpp

```diff
@@ -291,44 +313,100 @@ void DiscoveryManager::HandleNodeIdResolve(void * context, MdnsService * result,
-            {
-                deliminatorFound = true;
-                break;
-            }
-            else
-            {
-                uint8_t val = HexToInt(result->mName[i]);
-
-                if (val == UINT8_MAX)
-                {
-                    break;
-                }
-                else
-                {
-                    nodeId = nodeId * 16 + val;
-                }
-            }
+            ChipLogProgress(Discovery, "Invalid service entry from node %" PRIX64, nodeId);
+            mgr->mResolveDelegate->HandleNodeIdResolve(error, kUndefinedNodeId, *result);
         }
-
-        if (deliminatorFound)
+        else
         {
             ChipLogProgress(Discovery, "Node ID resolved for %" PRIX64, nodeId);
             mgr->mResolveDelegate->HandleNodeIdResolve(error, nodeId, *result);
         }
+    }
+#endif // CHIP_ENABLE_MDNS
+}
+
+#if CHIP_ENABLE_MDNS
+CHIP_ERROR DiscoveryManager::BrowseUnprovisionedDevice(Inet::IPAddressType type, Inet::InterfaceId interface)
+{
+    return ChipMdnsBrowse("_chipc", MdnsServiceProtocol::kMdnsProtocolUdp, type, interface, HandleUnprovisionedDeviceBrowse, this);
+}
+
+void DiscoveryManager::HandleUnprovisionedDeviceBrowse(void * context, MdnsService * services, size_t servicesSize,
+                                                       CHIP_ERROR error)
+{
+    DiscoveryManager * mgr = static_cast<DiscoveryManager *>(context);
+
+    if (mgr->mResolveDelegate)
+    {
+        mgr->mResolveDelegate->HandleUnprovisionedDeviceBrowse(error, services, servicesSize);
+    }
+}
+
+CHIP_ERROR DiscoveryManager::ResolveUnprovisionedDevice(uint64_t name, Inet::IPAddressType type, Inet::InterfaceId interface)
+{
+    MdnsService service;
+
+    snprintf(service.mName, sizeof(service.mName), "%016" PRIX64, name);
+    strncpy(service.mType, "_chipc", sizeof(service.mType));
+    service.mProtocol    = MdnsServiceProtocol::kMdnsProtocolUdp;
+    service.mAddressType = type;
+    return ChipMdnsResolve(&service, interface, HandleDiscriminatorResolve, this);
+}
+
+void DiscoveryManager::HandleDiscriminatorResolve(void * context, MdnsService * result, CHIP_ERROR error)
+{
+    DiscoveryManager * mgr = static_cast<DiscoveryManager *>(context);
+
+    if (mgr->mResolveDelegate == nullptr)
+    {
+        return;
+    }
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(Discovery, "Node ID resolved failed with %s", chip::ErrorStr(error));
+        mgr->mResolveDelegate->HandleUnprovisionedDeviceResolve(error, 0, MdnsService{});
+    }
+    else if (result == nullptr)
+    {
+        ChipLogError(Discovery, "Node ID resolve not found");
+        mgr->mResolveDelegate->HandleUnprovisionedDeviceResolve(CHIP_ERROR_UNKNOWN_RESOURCE_ID, 0, MdnsService{});
+    }
+    else
+    {
+        uint64_t name = 0;
+        char * end    = nullptr;
+
+        HexStringToUint64(result->mName, sizeof(result->mName), &end);
+
+        if (end == nullptr || *end != '\0')
```

Comment:
same here: I do not think end can be nullptr.

Reply:
this code doesn't cover another failure mode of `HexStringToUint64()`: when `end == result->mName`, which can occur when the empty string is in `result->mName`.  this code assumes that `result->mName` is always null-terminated.  is that always true?